experimental/python/wheel.bzl

```diff
@@ -0,0 +1,159 @@
+
+"""Rules for building wheels."""
+
+def _py_wheel_impl(ctx):
+    outfile = ctx.actions.declare_file("-".join([
+        ctx.attr.wheel_name,
+        ctx.attr.version,
+        ctx.attr.python,
+        ctx.attr.abi,
+        ctx.attr.platform,
+    ]) + ".whl")
+
+    inputs = depset(
+        transitive = [dep[DefaultInfo].data_runfiles.files for dep in ctx.attr.deps] +
+                     [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps],
+    )
+    arguments = [
+        "--name",
+        ctx.attr.wheel_name,
+        "--version",
+        ctx.attr.version,
+        "--python",
+        ctx.attr.python,
+        "--abi",
+        ctx.attr.abi,
+        "--platform",
+        ctx.attr.platform,
+        "--out",
+        outfile.path,
+    ]
+    arguments.extend(["--restrict_package=%s" % p for p in ctx.attr.packages])
+    for input_file in inputs.to_list():
+        arguments.append("--input_file")
+        arguments.append("%s;%s" % (input_file.short_path, input_file.path))
+
+    extra_headers = []
+    if ctx.attr.author:
+        extra_headers.append("Author: %s" % ctx.attr.author)
+    if ctx.attr.author_email:
+        extra_headers.append("Author-email: %s" % ctx.attr.author_email)
+    if ctx.attr.homepage:
+        extra_headers.append("Home-page: %s" % ctx.attr.homepage)
+    if ctx.attr.license:
+        extra_headers.append("License: %s" % ctx.attr.license)
+
+    for h in extra_headers:
+        arguments.append("--header")
+        arguments.append(h)
+
+    for c in ctx.attr.classifiers:
+        arguments.append("--classifier")
+        arguments.append(c)
+
+    for r in ctx.attr.requires:
+        arguments.append("--requires")
+        arguments.append(r)
+
+    for option, requirements in ctx.attr.extra_requires.items():
+        for r in requirements:
+            arguments.append("--extra_requires")
+            arguments.append(r + ";" + option)
+
+    for name, ref in ctx.attr.console_scripts.items():
+        arguments.append("--console_script")
+        arguments.append(name + " = " + ref)
+
+    if ctx.attr.description_file:
+        description_files = ctx.attr.description_file.files.to_list()
+        arguments.append("--description_file")
+        arguments.append(description_files[0].path)
+        inputs = inputs.union(ctx.attr.description_file.files)
+
+    ctx.actions.run(
+        inputs = inputs,
+        outputs = [outfile],
+        arguments = arguments,
+        executable = ctx.executable._wheelmaker,
+        progress_message = "Building wheel",
+    )
+    return [DefaultInfo(
+        files = depset([outfile]),
+        data_runfiles = ctx.runfiles(files = [outfile]),
+    )]
+
+py_wheel = rule(
+    implementation = _py_wheel_impl,
+    attrs = {
+        "deps": attr.label_list(
+            doc = "Dependencies of this package, e.g. py_library rules",
+        ),
+        "packages": attr.string_list(
+            mandatory = True,
+            allow_empty = False,
```

Comment:
This is required and can't be empty, yet `file_matches_packages` does support an empty filter list (in which case all files are emitted).

What if I want to produce a wheel that contains the entire content of my `py_library` within my repo? Or even my whole workspace?

Also, does the package path account for the repo name? E.g. 

```
experimental.examples.wheel.lib.simple_module
```
vs
```
io_bazel_rules_python.experimental.examples.wheel.lib.simple_module
```

We're unfortunately ambiguous about which of those is canonical, but we want to move to the latter. See bazelbuild/bazel#7067.

Reply:
In the runfiles tree, the repo name is already included underneath the runfiles root. So hopefully the wheel layout is the same as the layout seen by a py_binary executed by Bazel.

Reply:
This hits a difference between actual runfiles tree and what's presented by runfiles object available to skylark.

In the runfiles object, the files have a root and the path. I'm stripping the root from the name before packaging  (_path_inside_wheel function); the root contains full filesystem path; moreover, all generated files are in separate roots while they shoould be merged into common structure. There's also short_path which is supposed to be the path without root, but in some cases it's relative (../pip_import/).